source/Calamari.Aws/Integration/CloudFormation/CloudFormationObjectExtensions.cs

```diff
@@ -141,6 +142,64 @@ public static Task<CreateChangeSetResponse> CreateChangeSetAsync(this Func<IAmaz
             }
         }
 
+        /// <summary>
+        /// Gets all stack events, optionally filtered by a predicate
+        /// </summary>
+        /// <param name="predicate">The optional predicate used to filter events</param>
+        /// <returns>The stack events</returns>
+        public static async Task<List<Maybe<StackEvent>>> GetStackEvents(this Func<IAmazonCloudFormation> clientFactory,
+            StackArn stack,
+            Func<StackEvent, bool> predicate = null)
+        {
+            try
+            {
+                var stackEventResults = new List<Maybe<StackEvent>>();
+
+                var response = await clientFactory().DescribeStackEventsAsync(new DescribeStackEventsRequest { StackName = stack.Value });
+
+                var stackEvents = response?
+                    .StackEvents.Where(stackEvent => predicate == null || predicate(stackEvent))                    
+                    .ToList();
+
+                stackEventResults.AddRange(stackEvents.Select(s => s.AsSome()));
+
+                var nestedStackIds = stackEvents
+                    .Where(s => s.ResourceType == "AWS::CloudFormation::Stack" && s.PhysicalResourceId != s.StackId)
+                    .Select(s => s.PhysicalResourceId)
+                    .Distinct()
+                    .ToList();
+
+                foreach (var nestedStackId in nestedStackIds)
+                {
+                    var nestedStackEvents = await GetStackEvents(clientFactory, new StackArn(nestedStackId), predicate);
+                    if (nestedStackEvents.Any())
+                    {
+                        stackEventResults.AddRange(nestedStackEvents);
+                    }
+                }
+
+                return stackEventResults.OrderBy(s => s.SelectValueOr(e => e.Timestamp, DateTime.MinValue)).ToList();
+            }
+            catch (AmazonCloudFormationException ex) when (ex.ErrorCode == "AccessDenied")
+            {
+                throw new PermissionException(
+                    "The AWS account used to perform the operation does not have the required permissions to query the current state of the CloudFormation stack. " +
+                    "This step will complete without waiting for the stack to complete, and will not fail if the stack finishes in an error state.\n " +
```

Comment:
Why `\n`?

Reply:
No particular reason, the code was taken from some existing code that interacts with CloudFormation API. Looks like quite a few error messages and exception handling messages are deliberately including \n maybe so that the output in the logs is a bit easier to follow as it can be quite verbose?

Reply:
I think this is something we should clean up. I really don't see our job formatting errors at this level.

Reply:
This is done